### ISSUE TITLE
[css-scroll-snap-2] Standardize snap point selection with multiple aligned boxes

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#9622]
The proposed [snap events](https://drafts.csswg.org/css-scroll-snap-2/#snap-events) and [styling](https://drafts.csswg.org/css-scroll-snap-2/#snapped) in css-scroll-snap-2 expose the concept of snapped targets. Today, the developer can tell which element a browser is "snapped to" by moving them and seeing which one [re-snapping after layout changes](https://drafts.csswg.org/css-scroll-snap-1/#re-snap) follows. The spec says the following about multiple snap aligned elements:

> If multiple boxes were snapped before and their snap positions no longer coincide, then if one of them is focused or targeted, the scroll container must re-snap to that one and otherwise which one to re-snap to is UA-defined. (The UA may, for example, track which element is snapped as layout shifts align and de-align the snap positions of other elements.)

TLDR,
* Outside of focus or a target, the specific snapped element is not defined.
* When you have focus and target it's not defined which one takes precedence.
* Doesn't say which axis alignment to prefer if there is a different target in each axis.

Bringing this back to exposing the snap target, we have a two options:
1. Treat all elements which are visually aligned as snapped.
2. Standardize and expose the actually selected snap area in each axis.

Option 1 is relatively simple to understand, though requires extra work to compute (as it's not what's tracked internally) and may not be very useful. E.g. imagine a 2d grid of elements with `snap-align: center`, every element along the horizontal and vertical line through the center is *snapped*.

Option 2 would be more efficient to implement, as it is what the browser must already track internally, and if we are able to standardize on the snap point selection with multiple currently snap aligned elements would allow developers to know the element which will be tracked through layout shifts.

I propose we go with option 2, and update [css-scroll-snap-1 4.1.3. Re-snapping After Layout Changes](https://drafts.csswg.org/css-scroll-snap-1/#re-snap) to always define an exact target in each axis. My proposal would be the following:

1. Let inline be the set of boxes currently aligned on the inline axis.
2. Let block be the set of boxes currently aligned on the block axis.
3. If there is a focused element in either / both lists, remove all other elements from the list.
4. If there is a targeted element in either / both lists, remove all other elements from the list.
5. Remove any items from the block and inline lists which contain (nest) inner items in those lists. I.e. prefer snapping to innermost elements when nesting snap areas.
6. If inline and block have elements in common, filter both lists to those elements they have in common.
7. Select the first element in [tree order](https://dom.spec.whatwg.org/#concept-tree-order) from each of the block and inline list as the snapped target in each axis.
8. When re-snapping, prefer to follow the axis that has the focused element, then the one which has the targeted element, then the block axis.

**Comment by argyleink** [comment]
Agree, **option 2.** It's the interface I'm lookin for in my snap experiences

**Comment by fantasai** [comment]
I think your proposal seems to make sense. :) The one bit I might change is to maybe flip the order of 5 & 6. (I'm not entirely sure about 5, I haven't thought through it thoroughly, but it seems to make sense.)

Wrt the spec, my suggestion would be to adopt your suggestion to prioritize focused over targetted, and otherwise inline your suggested behavior as a non-normative example. And then after we've got some experience with it and are feeling confident it's the right algorithm, encode it normatively in the spec.

**Comment by flackr** [comment]
> I think your proposal seems to make sense. :)

Thanks for reviewing!

> The one bit I might change is to maybe flip the order of 5 & 6.

Yes, I debated the order here as well. I think that 5 makes sense to happen first but haven't worked through specific cases where they'd produce different answers yet. 

> (I'm not entirely sure about 5, I haven't thought through it thoroughly, but it seems to make sense.)

If you have a large snap area, you want to prefer the inner item that's snapped to. This should play well with the algorithm we define for nested snap point selection in https://github.com/w3c/csswg-drafts/issues/9187 which tries to clearly snap to inner items.

E.g. consider when you're in a valid snap position for inner and outer below:
```
 -----------
|   outer   |
|           |
|           |
|           |
|           |
|  -------  |
| | inner | |
| |       | |
| |       | |
|  -------  |
|           |
|           |
|           |
 -----------
```

If inner is in a valid snap position then it is a more reasonable current target than outer.
